cloudinit/cmd/main.py

```diff
@@ -461,133 +462,150 @@ def main_init(name, args):
     # Any log usage prior to setup_logging above did not have local user log
     # config applied.  We send the welcome message now, as stderr/out have
     # been redirected and log now configured.
-    welcome(name, msg=w_msg)
-    LOG.info("PID [%s] started cloud-init '%s'.", os.getppid(), bootstage_name)
+    welcome(stage.welcome)
+    LOG.info("PID [%s] started cloud-init '%s'.", os.getppid(), stage.welcome)
 
     # re-play early log messages before logging was setup
     for lvl, msg in early_logs:
         LOG.log(lvl, msg)
 
     # Stage 3
     try:
-        init.initialize()
+        init.initialize_filesystem()
     except Exception:
         util.logexc(LOG, "Failed to initialize, likely bad things to come!")
     # Stage 4
-    path_helper = init.paths
-    purge_cache_on_python_version_change(init)
-    mode = sources.DSMODE_LOCAL if args.local else sources.DSMODE_NETWORK
+    init.start_boot()
+    init.track_python_version_change()
 
-    if mode == sources.DSMODE_NETWORK:
```

Comment:
Using dsmode variables for flow control that isn't actually relevant to the dsmode is unnecessarily convoluted.

All stage-specific flow control* now uses the local `stage` variable and the `stages.{local,network,config,final}` dataclasses.

*with the exception of the (two?) spots in the code where datasource.dsmode modifies the flow of code